File: 357/solution.js
main();

function main() {
  expect(["Goku", "Vegeta", "Piccolo", "Gohan"], ["Piccolo", "Vegeta", "Gohan", "Goku"]);

  // FIXME: Expected array should have the positions of "Winry" and "Roy" swapped.
  expect(["Edward", "Alphonse", "Roy", "Winry"], ["Alphonse", "Edward", "Winry", "Roy"]);
}

function expect(names, expected) {
  const actual = sortNames(names);
  const exactMatch = actual.every((_, i) => actual[i] == expected[i]);
  if (!exactMatch) {
    console.error(`Expected sortNames([${names}]) == [${expected}], got [${actual}]`);
  }
}

function sortNames(names) {
  const collator = new Intl.Collator("en");

  return names.toSorted((nameA, nameB) => {
    const vowelsInA = countVowels(nameA);
    const vowelsInB = countVowels(nameB);
    const diff = vowelsInB - vowelsInA;
    if (diff == 0) {
      return collator.compare(nameA, nameB);
    }
    return diff;
  });
}

function countVowels(name) {
  return name.match(/[aeiou]/gi)?.length ?? 0;
}

File: 357/solution.go
package main

import (
	"cmp"
	"fmt"
	"regexp"
	"slices"
)

var vowelRegexp = regexp.MustCompile(`(?i)[aeiou]`)

func main() {
	expect([]string{"Goku", "Vegeta", "Piccolo", "Gohan"}, []string{"Piccolo", "Vegeta", "Gohan", "Goku"})

	// FIXME: Expected array should have the positions of "Winry" and "Roy" swapped.
	expect([]string{"Edward", "Alphonse", "Roy", "Winry"}, []string{"Alphonse", "Edward", "Winry", "Roy"})
}

func expect(names, expected []string) {
	actual := sortNames(names)

	exactMatch := sliceEvery(actual, func(_ string, i int) bool {
		return actual[i] == expected[i]
	})

	if !exactMatch {
		fmt.Printf("Expected sortNames(%v) == %v, got %v\n", names, expected, actual)
	}
}

func sortNames(names []string) []string {
	clonedNames := slices.Clone(names)
	slices.SortFunc(clonedNames, func(nameA, nameB string) int {
		if d := countVowels(nameB) - countVowels(nameA); d != 0 {
			return d
		}
		return cmp.Compare(nameA, nameB)
	})
	return clonedNames
}

func countVowels(name string) int {
	return len(vowelRegexp.FindAllString(name, len(name)))
}

func sliceEvery[T any](sl []T, f func(T, int) bool) bool {
	for i := range sl {
		if !f(sl[i], i) {
			return false
		}
	}
	return true
}
